// @sapphire/pieces/dist/cjs/lib/structures/Piece.cjs
'use strict';

var Container_cjs = require('../shared/Container.cjs');
var PieceLocation_cjs = require('./PieceLocation.cjs');

var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __publicField = (obj, key, value) => {
  __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
  return value;
};
var _Piece = class _Piece {
  constructor(context, options = {}) {
    /**
     * The store that contains the piece.
     */
    __publicField(this, "store");
    /**
     * The location metadata for the piece's file.
     */
    __publicField(this, "location");
    /**
     * The name of the piece.
     */
    __publicField(this, "name");
    /**
     * Whether or not the piece is enabled.
     */
    __publicField(this, "enabled");
    /**
     * The raw options passed to this {@link Piece}
     */
    __publicField(this, "options");
    this.store = context.store;
    this.location = new PieceLocation_cjs.PieceLocation(context.path, context.root);
    this.name = options.name ?? context.name;
    this.enabled = options.enabled ?? true;
    this.options = options;
  }
  /**
   * A reference to the {@link Container} object for ease of use.
   * @see container
   */
  get container() {
    return Container_cjs.container;
  }
  /**
   * Per-piece listener that is called when the piece is loaded into the store.
   * Useful to set-up asynchronous initialization tasks.
   */
  onLoad() {
    return void 0;
  }
  /**
   * Per-piece listener that is called when the piece is unloaded from the store.
   * Useful to set-up clean-up tasks.
   */
  onUnload() {
    return void 0;
  }
  /**
   * Unloads and disables the piece.
   */
  async unload() {
    await this.store.unload(this.name);
    this.enabled = false;
  }
  /**
   * Reloads the piece by loading the same path in the store.
   */
  async reload() {
    await this.store.load(this.location.root, this.location.relative);
  }
  /**
   * Defines the `JSON.stringify` behavior of this piece.
   */
  toJSON() {
    return {
      location: this.location.toJSON(),
      name: this.name,
      enabled: this.enabled,
      options: this.options
    };
  }
};
__name(_Piece, "Piece");
exports.Piece = _Piece;
((Piece2) => {
  Piece2.Location = PieceLocation_cjs.PieceLocation;
})(exports.Piece || (exports.Piece = {}));
