// @sapphire/framework/dist/esm/preconditions/GuildOnly.mjs
import { __name } from '../chunk-G5GHKT7C.mjs';
import { container } from '@sapphire/pieces';
import { Identifiers } from '../lib/errors/Identifiers.mjs';
import { AllFlowsPrecondition } from '../lib/structures/Precondition.mjs';

var _CorePrecondition = class _CorePrecondition extends AllFlowsPrecondition {
  messageRun(message) {
    return message.guildId === null ? this.makeSharedError() : this.ok();
  }
  chatInputRun(interaction) {
    return interaction.guildId === null ? this.makeSharedError() : this.ok();
  }
  contextMenuRun(interaction) {
    return interaction.guildId === null ? this.makeSharedError() : this.ok();
  }
  makeSharedError() {
    return this.error({
      // eslint-disable-next-line deprecation/deprecation
      identifier: Identifiers.PreconditionGuildOnly,
      message: "You cannot run this command in DMs."
    });
  }
};
__name(_CorePrecondition, "CorePrecondition");
var CorePrecondition = _CorePrecondition;
void container.stores.loadPiece({
  name: "GuildOnly",
  piece: CorePrecondition,
  store: "preconditions"
});

export { CorePrecondition };
