// @sapphire/pieces/dist/esm/lib/structures/PieceLocation.mjs
import { __name, __publicField } from '../../chunk-LIQV5WSN.mjs';
import { relative, sep, basename } from 'path';
import { VirtualPath } from '../internal/constants.mjs';

var _PieceLocation = class _PieceLocation {
  /**
   * @param full The full path to the file.
   * @param root The root directory the file was found from.
   */
  constructor(full, root) {
    /**
     * The full path to the file.
     */
    __publicField(this, "full");
    /**
     * The root directory the file was found from.
     */
    __publicField(this, "root");
    this.full = full;
    this.root = root;
  }
  /**
   * Whether the file is virtual or not.
   */
  get virtual() {
    return this.full === VirtualPath;
  }
  /**
   * The relative path between {@link PieceLocation.root} and {@link PieceLocation.full}.
   * @example
   * ```typescript
   * const location = new PieceLocation(
   * 	'/usr/src/app/commands',
   * 	'/usr/src/app/commands/general/ping.js'
   * );
   *
   * console.log(location.relative);
   * // → 'general/ping.js'
   * ```
   */
  get relative() {
    return this.virtual ? VirtualPath : relative(this.root, this.full);
  }
  /**
   * The names of the directories that separate {@link PieceLocation.root} and {@link PieceLocation.full}.
   * @example
   * ```typescript
   * const location = new PieceLocation(
   * 	'/usr/src/app/commands',
   * 	'/usr/src/app/commands/games/multiplayer/connect-four.js'
   * );
   *
   * console.log(location.directories);
   * // → ['games', 'multiplayer']
   * ```
   */
  get directories() {
    return this.virtual ? [] : this.relative.split(sep).slice(0, -1);
  }
  /**
   * The name and extension of the file that was loaded, extracted from {@link PieceLocation.full}.
   * @example
   * ```typescript
   * const location = new PieceLocation(
   * 	'/usr/src/app/commands',
   * 	'/usr/src/app/commands/games/multiplayer/connect-four.js'
   * );
   *
   * console.log(location.name);
   * // → 'connect-four.js'
   * ```
   */
  get name() {
    return this.virtual ? VirtualPath : basename(this.full);
  }
  /**
   * Defines the `JSON.stringify` behavior of this structure.
   */
  toJSON() {
    return {
      directories: this.directories,
      full: this.full,
      name: this.name,
      relative: this.relative,
      root: this.root
    };
  }
};
__name(_PieceLocation, "PieceLocation");
var PieceLocation = _PieceLocation;

export { PieceLocation };
