// @sapphire/framework/dist/esm/arguments/CoreDate.mjs
import { __name } from '../chunk-G5GHKT7C.mjs';
import { container } from '@sapphire/pieces';
import { Identifiers } from '../lib/errors/Identifiers.mjs';
import { resolveDate } from '../lib/resolvers/date.mjs';
import { Argument } from '../lib/structures/Argument.mjs';

var _CoreArgument = class _CoreArgument extends Argument {
  constructor(context) {
    super(context, { name: "date" });
    this.messages = {
      [Identifiers.ArgumentDateTooEarly]: ({ minimum }) => `The given date must be after ${new Date(minimum).toISOString()}.`,
      [Identifiers.ArgumentDateTooFar]: ({ maximum }) => `The given date must be before ${new Date(maximum).toISOString()}.`,
      [Identifiers.ArgumentDateError]: () => "The argument did not resolve to a date."
    };
  }
  run(parameter, context) {
    const resolved = resolveDate(parameter, { minimum: context.minimum, maximum: context.maximum });
    return resolved.mapErrInto(
      (identifier) => this.error({
        parameter,
        identifier,
        message: this.messages[identifier](context),
        context
      })
    );
  }
};
__name(_CoreArgument, "CoreArgument");
var CoreArgument = _CoreArgument;
void container.stores.loadPiece({
  name: "date",
  piece: CoreArgument,
  store: "arguments"
});

export { CoreArgument };
