// @sapphire/framework/dist/esm/arguments/CoreGuildTextChannel.mjs
import { __name } from '../chunk-G5GHKT7C.mjs';
import { container } from '@sapphire/pieces';
import { Identifiers } from '../lib/errors/Identifiers.mjs';
import { resolveGuildTextChannel } from '../lib/resolvers/guildTextChannel.mjs';
import { Argument } from '../lib/structures/Argument.mjs';

var _CoreArgument = class _CoreArgument extends Argument {
  constructor(context) {
    super(context, { name: "guildTextChannel" });
  }
  run(parameter, context) {
    const { guild } = context.message;
    if (!guild) {
      return this.error({
        parameter,
        identifier: Identifiers.ArgumentGuildChannelMissingGuildError,
        message: "This command can only be used in a server.",
        context
      });
    }
    const resolved = resolveGuildTextChannel(parameter, guild);
    return resolved.mapErrInto(
      (identifier) => this.error({
        parameter,
        identifier,
        message: "The given argument did not resolve to a valid text channel.",
        context: { ...context, guild }
      })
    );
  }
};
__name(_CoreArgument, "CoreArgument");
var CoreArgument = _CoreArgument;
void container.stores.loadPiece({
  name: "guildTextChannel",
  piece: CoreArgument,
  store: "arguments"
});

export { CoreArgument };
