// @sapphire/framework/dist/esm/arguments/CoreString.mjs
import { __name } from '../chunk-G5GHKT7C.mjs';
import { container } from '@sapphire/pieces';
import { Identifiers } from '../lib/errors/Identifiers.mjs';
import { resolveString } from '../lib/resolvers/string.mjs';
import { Argument } from '../lib/structures/Argument.mjs';

var _CoreArgument = class _CoreArgument extends Argument {
  constructor(context) {
    super(context, { name: "string" });
    this.messages = {
      [Identifiers.ArgumentStringTooShort]: ({ minimum }) => `The argument must be longer than ${minimum} characters.`,
      [Identifiers.ArgumentStringTooLong]: ({ maximum }) => `The argument must be shorter than ${maximum} characters.`
    };
  }
  run(parameter, context) {
    const resolved = resolveString(parameter, { minimum: context?.minimum, maximum: context?.maximum });
    return resolved.mapErrInto(
      (identifier) => this.error({
        parameter,
        identifier,
        message: this.messages[identifier](context),
        context
      })
    );
  }
};
__name(_CoreArgument, "CoreArgument");
var CoreArgument = _CoreArgument;
void container.stores.loadPiece({
  name: "string",
  piece: CoreArgument,
  store: "arguments"
});

export { CoreArgument };
