// @sapphire/framework/dist/esm/lib/utils/logger/Logger.mjs
import { __name } from '../../../chunk-G5GHKT7C.mjs';
import { LogLevel } from './ILogger.mjs';

var _Logger = class _Logger {
  constructor(level) {
    this.level = level;
  }
  has(level) {
    return level >= this.level;
  }
  trace(...values) {
    this.write(LogLevel.Trace, ...values);
  }
  debug(...values) {
    this.write(LogLevel.Debug, ...values);
  }
  info(...values) {
    this.write(LogLevel.Info, ...values);
  }
  warn(...values) {
    this.write(LogLevel.Warn, ...values);
  }
  error(...values) {
    this.write(LogLevel.Error, ...values);
  }
  fatal(...values) {
    this.write(LogLevel.Fatal, ...values);
  }
  write(level, ...values) {
    if (!this.has(level))
      return;
    const method = _Logger.levels.get(level);
    if (typeof method === "string")
      console[method](`[${method.toUpperCase()}]`, ...values);
  }
};
__name(_Logger, "Logger");
_Logger.levels = /* @__PURE__ */ new Map([
  [LogLevel.Trace, "trace"],
  [LogLevel.Debug, "debug"],
  [LogLevel.Info, "info"],
  [LogLevel.Warn, "warn"],
  [LogLevel.Error, "error"],
  [LogLevel.Fatal, "error"]
]);
var Logger = _Logger;

export { Logger };
