// @sapphire/framework/dist/cjs/arguments/CoreMember.cjs
'use strict';

var pieces = require('@sapphire/pieces');
var Identifiers_cjs = require('../lib/errors/Identifiers.cjs');
var member_cjs = require('../lib/resolvers/member.cjs');
var Argument_cjs = require('../lib/structures/Argument.cjs');

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var _CoreArgument = class _CoreArgument extends Argument_cjs.Argument {
  constructor(context) {
    super(context, { name: "member" });
  }
  async run(parameter, context) {
    const { guild } = context.message;
    if (!guild) {
      return this.error({
        parameter,
        identifier: Identifiers_cjs.Identifiers.ArgumentMemberMissingGuild,
        message: "This command can only be used in a server.",
        context
      });
    }
    const resolved = await member_cjs.resolveMember(parameter, guild, context.performFuzzySearch ?? true);
    return resolved.mapErrInto(
      (identifier) => this.error({
        parameter,
        identifier,
        message: "The given argument did not resolve to a server member.",
        context: { ...context, guild }
      })
    );
  }
};
__name(_CoreArgument, "CoreArgument");
var CoreArgument = _CoreArgument;
void pieces.container.stores.loadPiece({
  name: "member",
  piece: CoreArgument,
  store: "arguments"
});

exports.CoreArgument = CoreArgument;
