// @sapphire/framework/dist/cjs/arguments/CoreInteger.cjs
'use strict';

var pieces = require('@sapphire/pieces');
var Identifiers_cjs = require('../lib/errors/Identifiers.cjs');
var integer_cjs = require('../lib/resolvers/integer.cjs');
var Argument_cjs = require('../lib/structures/Argument.cjs');

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var _CoreArgument = class _CoreArgument extends Argument_cjs.Argument {
  constructor(context) {
    super(context, { name: "integer" });
    this.messages = {
      [Identifiers_cjs.Identifiers.ArgumentIntegerTooSmall]: ({ minimum }) => `The given number must be greater than ${minimum}.`,
      [Identifiers_cjs.Identifiers.ArgumentIntegerTooLarge]: ({ maximum }) => `The given number must be less than ${maximum}.`,
      [Identifiers_cjs.Identifiers.ArgumentIntegerError]: () => "The argument did not resolve to a valid number."
    };
  }
  run(parameter, context) {
    const resolved = integer_cjs.resolveInteger(parameter, { minimum: context.minimum, maximum: context.maximum });
    return resolved.mapErrInto(
      (identifier) => this.error({
        parameter,
        identifier,
        message: this.messages[identifier](context),
        context
      })
    );
  }
};
__name(_CoreArgument, "CoreArgument");
var CoreArgument = _CoreArgument;
void pieces.container.stores.loadPiece({
  name: "integer",
  piece: CoreArgument,
  store: "arguments"
});

exports.CoreArgument = CoreArgument;
