// @sapphire/framework/dist/cjs/listeners/application-commands/context-menu/CorePossibleContextMenuCommand.cjs
'use strict';

var pieces = require('@sapphire/pieces');
var Listener_cjs = require('../../../lib/structures/Listener.cjs');
var Events_cjs = require('../../../lib/types/Events.cjs');

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var _CoreListener = class _CoreListener extends Listener_cjs.Listener {
  constructor(context) {
    super(context, { event: Events_cjs.Events.PossibleContextMenuCommand });
  }
  run(interaction) {
    const { client, stores } = this.container;
    const commandStore = stores.get("commands");
    const command = commandStore.get(interaction.commandId) ?? commandStore.get(interaction.commandName);
    if (!command) {
      client.emit(Events_cjs.Events.UnknownContextMenuCommand, {
        interaction,
        context: { commandId: interaction.commandId, commandName: interaction.commandName }
      });
      return;
    }
    if (!command.contextMenuRun) {
      client.emit(Events_cjs.Events.CommandDoesNotHaveContextMenuCommandHandler, {
        command,
        interaction,
        context: { commandId: interaction.commandId, commandName: interaction.commandName }
      });
      return;
    }
    client.emit(Events_cjs.Events.PreContextMenuCommandRun, {
      command,
      context: { commandId: interaction.commandId, commandName: interaction.commandName },
      interaction
    });
  }
};
__name(_CoreListener, "CoreListener");
var CoreListener = _CoreListener;
void pieces.container.stores.loadPiece({
  name: "CorePossibleContextMenuCommand",
  piece: CoreListener,
  store: "listeners"
});

exports.CoreListener = CoreListener;
