// @sapphire/framework/dist/cjs/lib/utils/logger/Logger.cjs
'use strict';

var ILogger_cjs = require('./ILogger.cjs');

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var _Logger = class _Logger {
  constructor(level) {
    this.level = level;
  }
  has(level) {
    return level >= this.level;
  }
  trace(...values) {
    this.write(ILogger_cjs.LogLevel.Trace, ...values);
  }
  debug(...values) {
    this.write(ILogger_cjs.LogLevel.Debug, ...values);
  }
  info(...values) {
    this.write(ILogger_cjs.LogLevel.Info, ...values);
  }
  warn(...values) {
    this.write(ILogger_cjs.LogLevel.Warn, ...values);
  }
  error(...values) {
    this.write(ILogger_cjs.LogLevel.Error, ...values);
  }
  fatal(...values) {
    this.write(ILogger_cjs.LogLevel.Fatal, ...values);
  }
  write(level, ...values) {
    if (!this.has(level))
      return;
    const method = _Logger.levels.get(level);
    if (typeof method === "string")
      console[method](`[${method.toUpperCase()}]`, ...values);
  }
};
__name(_Logger, "Logger");
_Logger.levels = /* @__PURE__ */ new Map([
  [ILogger_cjs.LogLevel.Trace, "trace"],
  [ILogger_cjs.LogLevel.Debug, "debug"],
  [ILogger_cjs.LogLevel.Info, "info"],
  [ILogger_cjs.LogLevel.Warn, "warn"],
  [ILogger_cjs.LogLevel.Error, "error"],
  [ILogger_cjs.LogLevel.Fatal, "error"]
]);
var Logger = _Logger;

exports.Logger = Logger;
