// @sapphire/framework/dist/esm/lib/structures/Listener.mjs
import { __name } from '../../chunk-G5GHKT7C.mjs';
import { Piece } from '@sapphire/pieces';
import { Result } from '@sapphire/result';
import { Events } from '../types/Events.mjs';

var _Listener = class _Listener extends Piece {
  constructor(context, options = {}) {
    super(context, options);
    this.emitter = typeof options.emitter === "undefined" ? this.container.client : (typeof options.emitter === "string" ? Reflect.get(this.container.client, options.emitter) : options.emitter) ?? null;
    this.event = options.event ?? this.name;
    this.once = options.once ?? false;
    this._listener = this.emitter && this.event ? this.once ? this._runOnce.bind(this) : this._run.bind(this) : null;
    if (this.emitter === null || this._listener === null)
      this.enabled = false;
  }
  toJSON() {
    return {
      ...super.toJSON(),
      once: this.once,
      event: this.event
    };
  }
  async _run(...args) {
    const result = await Result.fromAsync(() => this.run(...args));
    result.inspectErr((error) => this.container.client.emit(Events.ListenerError, error, { piece: this }));
  }
  async _runOnce(...args) {
    await this._run(...args);
    await this.unload();
  }
};
__name(_Listener, "Listener");
var Listener = _Listener;

export { Listener };
