// @sapphire/framework/dist/cjs/lib/types/Enums.cjs
'use strict';

// src/lib/types/Enums.ts
var CooldownLevel = /* @__PURE__ */ ((CooldownLevel2) => {
  CooldownLevel2["Author"] = "author";
  CooldownLevel2["Channel"] = "channel";
  CooldownLevel2["Guild"] = "guild";
  return CooldownLevel2;
})(CooldownLevel || {});
var PluginHook = /* @__PURE__ */ ((PluginHook2) => {
  PluginHook2["PreGenericsInitialization"] = "preGenericsInitialization";
  PluginHook2["PreInitialization"] = "preInitialization";
  PluginHook2["PostInitialization"] = "postInitialization";
  PluginHook2["PreLogin"] = "preLogin";
  PluginHook2["PostLogin"] = "postLogin";
  return PluginHook2;
})(PluginHook || {});
var BucketScope = /* @__PURE__ */ ((BucketScope2) => {
  BucketScope2[BucketScope2["Channel"] = 0] = "Channel";
  BucketScope2[BucketScope2["Global"] = 1] = "Global";
  BucketScope2[BucketScope2["Guild"] = 2] = "Guild";
  BucketScope2[BucketScope2["User"] = 3] = "User";
  return BucketScope2;
})(BucketScope || {});
var RegisterBehavior = /* @__PURE__ */ ((RegisterBehavior2) => {
  RegisterBehavior2["Overwrite"] = "OVERWRITE";
  RegisterBehavior2["LogToConsole"] = "LOG_TO_CONSOLE";
  RegisterBehavior2["VerboseOverwrite"] = "VERBOSE_OVERWRITE";
  RegisterBehavior2["BulkOverwrite"] = "BULK_OVERWRITE";
  return RegisterBehavior2;
})(RegisterBehavior || {});
var InternalRegistryAPIType = /* @__PURE__ */ ((InternalRegistryAPIType2) => {
  InternalRegistryAPIType2[InternalRegistryAPIType2["ChatInput"] = 0] = "ChatInput";
  InternalRegistryAPIType2[InternalRegistryAPIType2["ContextMenu"] = 1] = "ContextMenu";
  return InternalRegistryAPIType2;
})(InternalRegistryAPIType || {});
var CommandOptionsRunTypeEnum = /* @__PURE__ */ ((CommandOptionsRunTypeEnum2) => {
  CommandOptionsRunTypeEnum2["Dm"] = "DM";
  CommandOptionsRunTypeEnum2["GuildText"] = "GUILD_TEXT";
  CommandOptionsRunTypeEnum2["GuildVoice"] = "GUILD_VOICE";
  CommandOptionsRunTypeEnum2["GuildNews"] = "GUILD_NEWS";
  CommandOptionsRunTypeEnum2["GuildNewsThread"] = "GUILD_NEWS_THREAD";
  CommandOptionsRunTypeEnum2["GuildPublicThread"] = "GUILD_PUBLIC_THREAD";
  CommandOptionsRunTypeEnum2["GuildPrivateThread"] = "GUILD_PRIVATE_THREAD";
  CommandOptionsRunTypeEnum2["GuildAny"] = "GUILD_ANY";
  return CommandOptionsRunTypeEnum2;
})(CommandOptionsRunTypeEnum || {});
var CommandPreConditions = /* @__PURE__ */ ((CommandPreConditions2) => {
  CommandPreConditions2["Cooldown"] = "Cooldown";
  CommandPreConditions2["DirectMessageOnly"] = "DMOnly";
  CommandPreConditions2["RunIn"] = "RunIn";
  CommandPreConditions2["GuildNewsOnly"] = "GuildNewsOnly";
  CommandPreConditions2["GuildNewsThreadOnly"] = "GuildNewsThreadOnly";
  CommandPreConditions2["GuildOnly"] = "GuildOnly";
  CommandPreConditions2["GuildPrivateThreadOnly"] = "GuildPrivateThreadOnly";
  CommandPreConditions2["GuildPublicThreadOnly"] = "GuildPublicThreadOnly";
  CommandPreConditions2["GuildTextOnly"] = "GuildTextOnly";
  CommandPreConditions2["GuildVoiceOnly"] = "GuildVoiceOnly";
  CommandPreConditions2["GuildThreadOnly"] = "GuildThreadOnly";
  CommandPreConditions2["NotSafeForWork"] = "NSFW";
  CommandPreConditions2["ClientPermissions"] = "ClientPermissions";
  CommandPreConditions2["UserPermissions"] = "UserPermissions";
  return CommandPreConditions2;
})(CommandPreConditions || {});

exports.BucketScope = BucketScope;
exports.CommandOptionsRunTypeEnum = CommandOptionsRunTypeEnum;
exports.CommandPreConditions = CommandPreConditions;
exports.CooldownLevel = CooldownLevel;
exports.InternalRegistryAPIType = InternalRegistryAPIType;
exports.PluginHook = PluginHook;
exports.RegisterBehavior = RegisterBehavior;
