// @sapphire/framework/dist/cjs/optional-listeners/message-command-listeners/CorePreMessageParser.cjs
'use strict';

var discord_jsUtilities = require('@sapphire/discord.js-utilities');
var discord_js = require('discord.js');
var Listener_cjs = require('../../lib/structures/Listener.cjs');
var Events_cjs = require('../../lib/types/Events.cjs');

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var _CoreListener = class _CoreListener extends Listener_cjs.Listener {
  constructor(context) {
    super(context, { event: Events_cjs.Events.PreMessageParsed });
    this.requiredPermissions = new discord_js.PermissionsBitField([discord_js.PermissionFlagsBits.ViewChannel, discord_js.PermissionFlagsBits.SendMessages]).freeze();
  }
  async run(message) {
    const canRun = await this.canRunInChannel(message);
    if (!canRun)
      return;
    let prefix = null;
    const mentionPrefix = this.getMentionPrefix(message);
    const { client } = this.container;
    const { regexPrefix } = client.options;
    if (mentionPrefix) {
      if (message.content.length === mentionPrefix.length) {
        client.emit(Events_cjs.Events.MentionPrefixOnly, message);
        return;
      }
      prefix = mentionPrefix;
    } else if (regexPrefix?.test(message.content)) {
      prefix = regexPrefix;
    } else {
      const prefixes = await client.fetchPrefix(message);
      const parsed = this.getPrefix(message.content, prefixes);
      if (parsed !== null)
        prefix = parsed;
    }
    if (prefix === null)
      client.emit(Events_cjs.Events.NonPrefixedMessage, message);
    else
      client.emit(Events_cjs.Events.PrefixedMessage, message, prefix);
  }
  async canRunInChannel(message) {
    if (discord_jsUtilities.isDMChannel(message.channel))
      return true;
    const me = await message.guild?.members.fetchMe();
    if (!me)
      return false;
    const { channel } = message;
    const permissionsFor = channel.permissionsFor(me);
    if (!permissionsFor)
      return false;
    return permissionsFor.has(this.requiredPermissions, true);
  }
  getMentionPrefix(message) {
    if (this.container.client.disableMentionPrefix)
      return null;
    if (message.content.length < 20 || !message.content.startsWith("<@"))
      return null;
    const [offset, id] = message.content[2] === "&" ? [3, message.guild?.roles.botRoleFor(this.container.client.id)?.id] : [message.content[2] === "!" ? 3 : 2, this.container.client.id];
    if (!id)
      return null;
    const offsetWithId = offset + id.length;
    if (message.content[offsetWithId] !== ">")
      return null;
    const mentionId = message.content.substring(offset, offsetWithId);
    if (mentionId === id)
      return message.content.substring(0, offsetWithId + 1);
    return null;
  }
  getPrefix(content, prefixes) {
    if (prefixes === null)
      return null;
    const { caseInsensitivePrefixes } = this.container.client.options;
    if (caseInsensitivePrefixes)
      content = content.toLowerCase();
    if (typeof prefixes === "string") {
      return content.startsWith(caseInsensitivePrefixes ? prefixes.toLowerCase() : prefixes) ? prefixes : null;
    }
    return prefixes.find((prefix) => content.startsWith(caseInsensitivePrefixes ? prefix.toLowerCase() : prefix)) ?? null;
  }
};
__name(_CoreListener, "CoreListener");
var CoreListener = _CoreListener;

exports.CoreListener = CoreListener;
